core/src/main/java/jenkins/security/apitoken/ApiTokenStore.java

```diff
@@ -0,0 +1,371 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.security.apitoken;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.Util;
+import hudson.util.Secret;
+import jenkins.security.Messages;
+import net.sf.json.JSONObject;
+import org.apache.commons.lang.StringUtils;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.UUID;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+public class ApiTokenStore {
+    private static final Logger LOGGER = Logger.getLogger(ApiTokenStore.class.getName());
+    private static final SecureRandom RANDOM = new SecureRandom();
```

Comment:
In Java 8, it may be preferable to use `SecureRandom.getInstanceStrong()` to prevent any bad SecureRandom implementations by being used.

Reply:
Concerning this point I was initially positively surprised to see a new method for that, but then when I use it, it throws a check exception, so the code becomes a lot longer (to fallback to default version in case of exception) and at that point I tried to investigate a bit the topic.

Here are different links ([issue tracker of Akka](https://github.com/akka/akka/issues/20908), [blog post](https://tersesystems.com/blog/2015/12/17/the-right-way-to-use-securerandom/), [stackoverflow](https://stackoverflow.com/a/27638413/2567174), [Synopsys blog post](https://www.synopsys.com/blogs/software-security/proper-use-of-javas-securerandom/)) that advice to not use this method due the blocking access of `/dev/random` without a positive impact on security. As Jenkins is often running inside container, the entropy of the system could not large enough.

So I let it like that for the moment except if you have arguments to change it.

Reply:
Yikes, I didn't realize it threw a checked exception! Sigh, the things you miss when you're in Scala or Kotlin land. :)

As for blocking on `/dev/random`, from what I've read, that hasn't been an actual problem in years. IIRC, Linux seeds this on boot which is like the only time it blocks, and then afterwards it's a PRNG just like `/dev/urandom`. Not sure how it's implemented in Windows, though, but I'm guessing it's similar.